sycl/include/CL/sycl/detail/common.hpp

```diff
@@ -113,6 +113,94 @@ const cl_mem_fence_flags CLK_GLOBAL_MEM_FENCE  = 0x02;
 const cl_mem_fence_flags CLK_CHANNEL_MEM_FENCE = 0x04;
 #endif // __SYCL_DEVICE_ONLY__
 
+// Produces N-dimensional object of type T whose all components are initialized
+// to given integer value.
+template <int N, template <int> class T> struct InitializedVal {
+  template <int Val> static T<N> &&get();
+};
+
+// Specialization for a one-dimensional type.
+template <template <int> class T> struct InitializedVal<1, T> {
+  template <int Val> static T<1> &&get() { return T<1>{Val}; }
+};
+
+// Specialization for a two-dimensional type.
+template <template <int> class T> struct InitializedVal<2, T> {
+  template <int Val> static T<2> &&get() { return T<2>{Val, Val}; }
+};
+
+// Specialization for a three-dimensional type.
+template <template <int> class T> struct InitializedVal<3, T> {
+  template <int Val> static T<3> &&get() { return T<3>{Val, Val, Val}; }
+};
+
+// Fills the lack of enable_if_t in C++11.
+template <bool Cond, typename T = void>
+using enable_if_t = typename std::enable_if<Cond, T>::type;
+
+// Generates an NDIMS-dimensional perfect loop nest. The purpose of this class
+// is to better support handling of situations where there must be a loop nest
+// over a multi-dimensional space - it allows to avoid generating unnecessary
+// outer loops like 'for (int z=0; z<1; z++)' in case of 1D and 2D iteration
+// spaces or writing specializations of the algorighms for 1D, 2D and 3D cases.
```

Comment:
Not particularly relevant as it's understandable either way but "algorighms" -> "algorithms"

Reply:
Thanks for the catch. Will change